File: api/analyse.js
const gplay = require('google-play-scraper');

const STOPWORDS = new Set([
  'the','a','an','and','or','but','in','on','at','to','for','of','with',
  'is','it','its','this','that','was','are','be','been','have','has','had',
  'do','does','did','will','would','could','should','not','no','so','if',
  'i','my','me','we','you','your','they','them','app','use','get','just',
  'also','very','good','great','nice','bad','even','still','after','before',
  'when','than','more','all','some','one','time','way','make','now','new',
  'well','really','much','many','like','please','thank','thanks','this',
]);

function extractPhrases(reviews, starRange, topN = 10) {
  const counts = {};
  reviews.filter(r => starRange.includes(r.score)).forEach(r => {
    const text = (r.text || r.content || '').toLowerCase().replace(/[^a-z\s]/g, ' ');
    const tokens = text.split(/\s+/).filter(t => t.length > 3 && !STOPWORDS.has(t));
    tokens.forEach(t => { counts[t] = (counts[t] || 0) + 1; });
    for (let i = 0; i < tokens.length - 1; i++) {
      const phrase = tokens[i] + ' ' + tokens[i+1];
      counts[phrase] = (counts[phrase] || 0) + 1;
    }
  });
  return Object.entries(counts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, topN)
    .map(([phrase, count]) => ({ phrase, count }));
}


// ... extractPhrases function above ...

function buildPeriodBreakdown(reviews) {
  const now = Date.now();
  const PERIODS = [7, 30, 60, 90, 120, 180, 270, 365];

  const getTime = (val) => {
    if (!val) return null;
    const t = val instanceof Date ? val.getTime() : new Date(val).getTime();
    return isNaN(t) ? null : t;
  };

  const timestamps = reviews.map(r => getTime(r.date)).filter(Boolean);
  if (!timestamps.length) return {};

  const oldestDate = Math.min(...timestamps);
  const actualDays = Math.ceil((now - oldestDate) / (86400 * 1000));
  const fmt = ts => new Date(ts).toLocaleDateString('en-IN', { day:'numeric', month:'short', year:'numeric' });

  const result = {};

  PERIODS.filter(d => d <= actualDays + 7).forEach(days => {
    const cutoff = now - days * 86400 * 1000;
    const subset = reviews.filter(r => { const t = getTime(r.date); return t && t >= cutoff; });
    if (!subset.length) return;
    const stars = {1:0,2:0,3:0,4:0,5:0};
    subset.forEach(r => { if (stars[r.score] !== undefined) stars[r.score]++; });
    const avg = subset.reduce((s, r) => s + r.score, 0) / subset.length;
    result[`Last ${days} Days`] = {
      days,
      dateRange: `${fmt(cutoff)} → ${fmt(now)}`,
      total: subset.length,
      avg: Math.round(avg * 100) / 100,
      '5star': stars[5], '4star': stars[4],
      '3star': stars[3], '2star': stars[2], '1star': stars[1],
    };
  });

  if (reviews.length) {
    const stars = {1:0,2:0,3:0,4:0,5:0};
    reviews.forEach(r => { if (stars[r.score] !== undefined) stars[r.score]++; });
    const avg = reviews.reduce((s, r) => s + r.score, 0) / reviews.length;
    result[`All Fetched (${reviews.length} reviews)`] = {
      days: null,
      dateRange: `${fmt(oldestDate)} → ${fmt(now)}`,
      total: reviews.length,
      avg: Math.round(avg * 100) / 100,
      '5star': stars[5], '4star': stars[4],
      '3star': stars[3], '2star': stars[2], '1star': stars[1],
    };
  }

  return result;
}



module.exports = async (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Content-Type', 'application/json');

  if (req.method === 'OPTIONS') return res.status(200).end();

  const { appId, count = '150' } = req.query;
  if (!appId) return res.status(400).json({ error: 'appId is required' });

  const numCount = Math.min(parseInt(count) || 150, 200000);

  try {
    // App info
    let info = {};
    try { info = await gplay.app({ appId, lang: 'en', country: 'in' }); } catch(_) {}

    // Fetch reviews in batches
    let allReviews = [], token = null;
    while (allReviews.length < numCount) {
      const batch = Math.min(200, numCount - allReviews.length);
      const reviewResult = await gplay.reviews({
        appId, lang: 'en', country: 'in',
        sort: gplay.sort.NEWEST,
        count: batch,
        continuationToken: token,
      });
    const result = reviewResult.data;
    token = reviewResult.nextPaginationToken;
      if (!result || !result.length) break;
      allReviews = allReviews.concat(result);
      if (!token) break;
    }

    if (!allReviews.length) {
      return res.status(404).json({ error: `No reviews found for ${appId}` });
    }



  // ── Map reviews first so all fields are correct ──
    const mappedReviews = allReviews.map(r => ({
      userName:  r.userName || 'Anonymous',
      score:     r.score,
      text:      r.content || r.text || '',
      date:      r.at ? (r.at instanceof Date ? r.at : new Date(r.at)).toISOString() : null,
      thumbsUp:  r.thumbsUpCount || 0,
      version:   r.reviewCreatedVersion || '',
      replyText: r.replyContent || '',
      replyDate: r.repliedAt ? (r.repliedAt instanceof Date ? r.repliedAt : new Date(r.repliedAt)).toISOString() : '',
    }));

    // ── Run analysis on mapped reviews (so r.text is correct) ──
    const topIssues    = extractPhrases(mappedReviews, [1, 2], 10);
    const topGood      = extractPhrases(mappedReviews, [4, 5], 10);
    const issueSummary = topIssues.length ? `Users most commonly mention: ${topIssues.slice(0,5).map(x => x.phrase).join(', ')}. These appear repeatedly in 1★–2★ reviews.` : null;
    const goodSummary  = topGood.length   ? `Users frequently praise: ${topGood.slice(0,5).map(x => x.phrase).join(', ')}. These themes dominate 4★–5★ reviews.` : null;

    const pos   = mappedReviews.filter(r => r.score >= 4).length;
    const neg   = mappedReviews.filter(r => r.score <= 2).length;
    const total = mappedReviews.length;

    const sentiment = pos / total > 0.6 ? 'positive'
                    : pos / total < 0.4 ? 'negative'
                    : 'mixed';

    return res.status(200).json({
      appId,
      appName:      info.title   || appId,
      appIcon:      info.icon    || null,
      appRating:    info.score   || null,
      totalReviews: info.ratings || info.reviews || null,
      analysed:     total,
      positive:     pos,
      negative:     neg,
      neutral:      total - pos - neg,
      sentiment,
      topIssues,
      topGood,
      issueSummary,
      goodSummary,
      periodBreakdown: buildPeriodBreakdown(mappedReviews),
      reviews: mappedReviews,
    });
    


  } catch (err) {
    return res.status(500).json({ error: err.message });
  }
};

File: api/analyse.py
from google_play_scraper import app as gplay_app, reviews, Sort
from collections import Counter
from datetime import datetime, timedelta
import json

ISSUE_KEYWORDS = {
    "App crashes / freezes":    ["crash","freeze","frozen","hang","stuck","not opening","force close","black screen"],
    "Login / OTP problems":     ["otp","login","sign in","can't log","logout","logged out","verify"],
    "Payment failures":         ["payment","failed","deducted","charged","refund","money","transaction"],
    "Slow / laggy":             ["slow","lag","lagging","loading","takes forever","buffer"],
    "Order / delivery issues":  ["order","deliver","late","delayed","wrong product","return"],
    "Customer support":         ["support","customer care","no response","ignored","useless","not helpful"],
    "Bugs / errors":            ["bug","error","glitch","broken","not working","doesn't work"],
    "Too many ads":             ["ad","ads","advertisement","popup","spam","notification"],
    "Poor UI / UX":             ["confusing","complicated","bad design","ugly","layout"],
}

GOOD_KEYWORDS = {
    "Easy to use":              ["easy","simple","smooth","user friendly","intuitive","clean"],
    "Fast & reliable":          ["fast","quick","instant","reliable","stable","works great"],
    "Good offers / cashback":   ["cashback","offer","discount","reward","deal","savings"],
    "Great customer support":   ["support","helpful","resolved","quick response","great service"],
    "Seamless payments":        ["seamless","easy payment","quick pay","smooth transaction"],
    "Trustworthy / secure":     ["safe","secure","trust","genuine","authentic"],
    "Value for money":          ["value","worth","affordable","cheap","cost effective"],
    "Overall love it":          ["love","amazing","excellent","fantastic","best app","perfect"],
}

def detect_themes(rv_list, star_range, keyword_bank):
    counts = Counter()
    for rv in rv_list:
        if rv.get("score") not in star_range:
            continue
        text = (rv.get("content") or "").lower()
        for theme, kws in keyword_bank.items():
            if any(kw in text for kw in kws):
                counts[theme] += 1
    return counts

def handler(request):
    # CORS
    headers = {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET, OPTIONS",
        "Content-Type": "application/json",
    }

    if request.method == "OPTIONS":
        return Response("", 200, headers)

    app_id = request.args.get("appId")
    count  = min(int(request.args.get("count", 200)), 500)

    if not app_id:
        return Response(json.dumps({"error": "appId required"}), 400, headers)

    try:
        info = gplay_app(app_id, lang="en", country="in")

        all_rv, token = [], None
        while len(all_rv) < count:
            batch = min(200, count - len(all_rv))
            result, token = reviews(
                app_id, lang="en", country="in",
                sort=Sort.NEWEST, count=batch,
                continuation_token=token
            )
            if not result: break
            all_rv.extend(result)
            if not token: break

        issue_counts = detect_themes(all_rv, {1, 2}, ISSUE_KEYWORDS)
        good_counts  = detect_themes(all_rv, {4, 5}, GOOD_KEYWORDS)

        pos = sum(1 for r in all_rv if r["score"] >= 4)
        neg = sum(1 for r in all_rv if r["score"] <= 2)
        total = len(all_rv)

        payload = {
            "appName":    info.get("title", app_id),
            "appIcon":    info.get("icon"),
            "appRating":  info.get("score"),
            "analysed":   total,
            "positive":   pos,
            "negative":   neg,
            "neutral":    total - pos - neg,
            "sentiment":  "positive" if pos/total > 0.6 else "negative" if pos/total < 0.4 else "mixed" if total else "mixed",
            "topIssues":  dict(issue_counts.most_common(8)),
            "topGood":    dict(good_counts.most_common(8)),
            "reviews": [
                {
                    "userName":  r.get("userName"),
                    "score":     r.get("score"),
                    "text":      r.get("content"),
                    "date":      r["at"].isoformat() if r.get("at") else None,
                    "thumbsUp":  r.get("thumbsUpCount", 0),
                    "version":   r.get("reviewCreatedVersion"),
                    "replyText": r.get("replyContent"),
                }
                for r in all_rv
            ],
        }

        return Response(json.dumps(payload), 200, headers)

    except Exception as e:
        return Response(json.dumps({"error": str(e)}), 500, headers)
```

### Step 2 — Create `requirements.txt` in your repo root
```
google-play-scraper==1.2.7
